File: app/src/main/java/com/artmurka/artmurkaapp/Presenter/CategoryPresenter.java
package com.artmurka.artmurkaapp.Presenter;

import android.content.Context;
import android.support.v4.app.FragmentManager;
import android.util.Log;

import com.artmurka.artmurkaapp.Model.Modules.AllRequestOvservers;
import com.artmurka.artmurkaapp.Model.InterfacesModel.IAllRequestObservers;
import com.artmurka.artmurkaapp.Model.Pojo.ItemList.Categories.Example;
import com.artmurka.artmurkaapp.Model.Pojo.ItemList.Categories.Success;
import com.artmurka.artmurkaapp.Other.SaveDataFragment;
import com.artmurka.artmurkaapp.Presenter.InterfacesPresenter.ICategoryPresenter;
import com.artmurka.artmurkaapp.Views.Fragments.Interfaces.ICategoryFragment;

import java.util.ArrayList;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;


public class CategoryPresenter implements ICategoryPresenter {

    private ICategoryFragment catFragment;
    private IAllRequestObservers model;
    private ArrayList<Success> successList;
    private Observable<Example> exampleObservable;
    private Context ctx;

    public CategoryPresenter(ICategoryFragment fragment, Context context) {
        catFragment = fragment;
        this.ctx = context;
        model = new AllRequestOvservers();
    }

    @Override
    public void getCategoriesData(final boolean isUpdate) {
        if (successList == null) {
            resetData();
        } else {
            catFragment.showCategories(successList);
        }
    }

    private void resetData() {

        exampleObservable = model.getCategories();
        exampleObservable.subscribe(new Observer<Example>() {
            @Override
            public void onSubscribe(Disposable d) {
            }

            @Override
            public void onNext(Example value) {
                successList = value.getSuccess();
                if (successList.isEmpty()) catFragment.showError("Error, succes list is Empty");
                else catFragment.showCategories(successList);
            }

            @Override
            public void onError(Throwable e) {
                catFragment.showError("Щось пішло не так. Перезавантажити ?");
                Log.d("Log.d", e.getMessage());
            }

            @Override
            public void onComplete() {
            }
        });
    }
}

File: app/src/main/java/com/artmurka/artmurkaapp/presenter/CategoryPresenter.kt
package com.artmurka.artmurkaapp.presenter

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.artmurka.artmurkaapp.R
import com.artmurka.artmurkaapp.data.model.pojo.itemlist.categories.Success
import com.artmurka.artmurkaapp.presenter.interfaces_presenter.ICategoryPresenter
import com.artmurka.artmurkaapp.android.views.fragments.interfaces.ICategoryFragment
import com.artmurka.artmurkaapp.domain.usecase.categories.GetCategoriesUseCase
import com.artmurka.artmurkaapp.other.Const
import java.util.ArrayList
import javax.inject.Inject
import io.reactivex.observers.DisposableObserver


class CategoryPresenter @Inject constructor(val model: GetCategoriesUseCase, val context: Context) : BasePresenter<ICategoryFragment>(), ICategoryPresenter {


    private var successList: ArrayList<Success>? = null

    fun makeCall() {
        val call = Uri.parse("tel:" + Const.TEL_NUMBER)
        val surf = Intent(Intent.ACTION_DIAL, call)
        surf.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context.startActivity(surf)
    }

    override fun getCategoriesData(isUpdate: Boolean) {
        if (successList == null) {
            resetData()
        } else {
            view?.showCategories(successList!!)
        }
    }

    private fun resetData() {
        model.execute(object : DisposableObserver<ArrayList<Success>>() {
            override fun onComplete() {}
            override fun onNext(t: ArrayList<Success>) {
                successList = t
                if (successList?.isEmpty()!!)
                    view?.showError("Error, succes list is Empty")
                else
                    view?.showCategories(successList!!)
            }

            override fun onError(e: Throwable) {
                view?.showError(context.resources.getString(R.string.any_error_reload))
            }
        }, GetCategoriesUseCase.Params())

    }

    override fun onDropView() {
        model.dispose()
        super.onDropView()
    }
}

